Ad7192 review no merge

mirceacaprioru opened the pull request:
This patch series is an effort to move the ad7192 out of staging. These commits try to solve some of the problems  refereed by Jonathan https://patchwork.kernel.org/patch/10741141/ 

 This modification provides a clock binding for clock reference, removes some platform data members and replaces them with dt bindings. These options will be subject to further changes and some of them will disappear entirely with the next patch series. 

This patch series is intended only for review purposes before sending upstream.

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -250,28 +253,37 @@ static int ad7192_setup(struct ad7192_state *st,
+						     "adi,int-clock-output-enable");
+		if (clock_out_en)
+			clock_sel = AD7192_CLK_INT_CO;
+	} else {
```
this if block could be it's own function;

commodo replied:
you can mask `-ENOENT` as a no-err case if moving into the func

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -250,28 +253,37 @@ static int ad7192_setup(struct ad7192_state *st,
+	st->fclk = AD7192_INT_FREQ_MHZ;
+
+	st->mclk = devm_clk_get(&st->sd.spi->dev, "clk");
+	if (IS_ERR(st->mclk)) {
```
for cases like `IS_ERR` & `-ENOENT`, it's a good idea to combine them in a single & exit early
smth like:
```
if (IS_ERR(st->mclk) && PTR_ERR(st->mclk) != -ENOENT)
    return PTR_ERR(st->mclk);
```

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -250,28 +253,37 @@ static int ad7192_setup(struct ad7192_state *st,
+		if (ret < 0)
+			return ret;
+
+		st->fclk = clk_get_rate(st->mclk);
```
for `fclk`, do you need to get the frequency only once on probe ?
or could this be made into a `ad7192_get_freq()` function ?

driver do this both ways; i don't think there's a single way to do it;
but it's worth asking

mirceacaprioru replied:
we only get this at driver probe. It does not modify after this point.

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -218,8 +218,8 @@ static const struct ad_sd_calib_data ad7192_calib_arr[8] = {
-		return ad_sd_calibrate_all(&st->sd, ad7192_calib_arr,
-				ARRAY_SIZE(ad7192_calib_arr));
+	return ad_sd_calibrate_all(&st->sd, ad7192_calib_arr,
+				   ARRAY_SIZE(ad7192_calib_arr));
```
this is nice;
it could be an idea to do a single white-space patch for the whole driver if doing just this;

commodo replied:
this patch can also be sent now as-is;
and not necessarily part of any series

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -229,23 +229,25 @@ static inline bool ad7192_valid_external_frequency(u32 freq)
 	ret = ad_sd_reset(&st->sd, 48);
 	if (ret < 0)
-		goto out;
+		return ret;
```
this should be moved into the other patch that handles just this `goto out  ->   return ret`  conversion

mirceacaprioru commented:
Changelog V1->V2

- clock setup moved to a separate function
- clock binding error verification simplified and return early if failure
- goto err moved to its own commit

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -226,10 +229,44 @@ static inline bool ad7192_valid_external_frequency(u32 freq)
+	st->fclk = AD7192_INT_FREQ_MHZ;
+
+	if (PTR_ERR(st->mclk) != -ENOENT) {
+		/* use internal clock */
```
this comment should be for the whole if block

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -226,10 +229,44 @@ static inline bool ad7192_valid_external_frequency(u32 freq)
+	if (PTR_ERR(st->mclk) != -ENOENT) {
+		/* use internal clock */
+		clock_out_en = of_property_read_bool(np,
+						     "adi,int-clock-output-enable");
```
`if (of_property_read_bool(np, "adi,int-clock-output-enable"))` and get rid of `clock_out_en ` var

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -226,10 +229,44 @@ static inline bool ad7192_valid_external_frequency(u32 freq)
+		/* use internal clock */
+		clock_out_en = of_property_read_bool(np,
+						     "adi,int-clock-output-enable");
+		if (clock_out_en)
```
same comment as above

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -226,10 +229,44 @@ static inline bool ad7192_valid_external_frequency(u32 freq)
+		if (clock_out_en)
+			st->clock_sel = AD7192_CLK_INT_CO;
+	} else {
+		clock_xtal = of_property_read_bool(np, "adi,clock-xtal");
```
same comment as above. This way you can also get rid of `clock_xtal ` var

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -226,10 +229,44 @@ static inline bool ad7192_valid_external_frequency(u32 freq)
+		else
+			st->clock_sel = AD7192_CLK_EXT_MCLK2;
+
+		ret = clk_prepare_enable(st->mclk);
```
you also need a `iio_device_unregister()`

mirceacaprioru commented:
Changelog V2 -> V3

- removed unnecessary bool varible
- modified clock name to mclk
- added clk_unprepare call

stefpopa commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -720,6 +741,7 @@ static int ad7192_probe(struct spi_device *spi)
 
 error_remove_trigger:
 	ad_sd_cleanup_buffer_and_trigger(indio_dev);
+	clk_unprepare(st->mclk);
```
you should call `clk_disable_unprepare(st->mclk)`. Also, you probably need another label called `error_clk_disable_unprepare` or something. You should goto this label if the `ad_sd_setup_buffer_and_trigger()` fails.

mirceacaprioru commented:
Changelog V4->V5

- add clk_disable_unprepare

mirceacaprioru commented:
Changelog v5->v6

- moved clk_prepare to probe to ensure the proper error sequence is being called
- changed ad7192_clock_select to ad7192_of_clock_select

commodo commented on drivers/staging/iio/adc/ad7192.c:
```diff
@@ -345,9 +345,6 @@ static int ad7192_setup(struct ad7192_state *st, struct device_node *np)
 	return 0;
-out:
-	dev_err(&st->sd.spi->dev, "setup failed\n");
-	return ret;
```
this patch looks good;
i think this can be sent now as-is;
and until we resolve others, it should get accepted

mirceacaprioru commented:
Should we merge this here if it has been upstreamed ?

commodo commented:
> Should we merge this here if it has been upstreamed ? It will help with future patches.

no; 
branch `iio-togreg-upstream` is jonathan's branch
it has no value for us if we merge it, since it will get overwritten
we only keep it for review purposes

let's close and push to master when all is ready
